pkg/virtualservice/resource_manager.go

```diff
@@ -0,0 +1,345 @@
+		return err
+	}
+
+	sdkVS, err := m.findSDKVirtualService(ctx, ms, vs)
+	if err != nil {
+		return err
+	}
+	if sdkVS == nil {
+		sdkVS, err = m.createSDKVirtualService(ctx, ms, vs, vnByKey, vrByKey)
+		if err != nil {
+			return err
+		}
+	} else {
+		sdkVS, err = m.updateSDKVirtualService(ctx, sdkVS, vs, vnByKey, vrByKey)
+		if err != nil {
+			return err
+		}
+	}
+	return m.updateCRDVirtualService(ctx, vs, sdkVS)
+}
+
+func (m *defaultResourceManager) Cleanup(ctx context.Context, vs *appmesh.VirtualService) error {
+	ms, err := m.findMeshDependency(ctx, vs)
+	if err != nil {
+		return err
+	}
+	if err := m.validateMeshDependencies(ctx, ms); err != nil {
+		return err
+	}
+	sdkVS, err := m.findSDKVirtualService(ctx, ms, vs)
+	if sdkVS == nil {
+		return nil
+	}
+	return m.deleteSDKVirtualService(ctx, sdkVS, vs)
+}
+
+// findMeshDependency find the Mesh dependency for this VirtualService.
+func (m *defaultResourceManager) findMeshDependency(ctx context.Context, vs *appmesh.VirtualService) (*appmesh.Mesh, error) {
+	if vs.Spec.MeshRef == nil {
+		return nil, errors.Errorf("meshRef shouldn't be nil, please check webhook setup")
+	}
+	ms, err := m.referencesResolver.ResolveMeshReference(ctx, *vs.Spec.MeshRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to resolve meshRef")
+	}
+	return ms, nil
+}
+
+// validateMeshDependencies validate the Mesh dependency for this VirtualService.
+func (m *defaultResourceManager) validateMeshDependencies(ctx context.Context, ms *appmesh.Mesh) error {
+	if !mesh.IsMeshActive(ms) {
+		return runtime.NewRequeueError(errors.New("mesh is not active yet"))
+	}
+	return nil
+}
+
+func (m *defaultResourceManager) findVirtualNodeDependencies(ctx context.Context, vs *appmesh.VirtualService) (map[types.NamespacedName]*appmesh.VirtualNode, error) {
+	if vs.Spec.Provider == nil || vs.Spec.Provider.VirtualNode == nil {
+		return nil, nil
+	}
+	vnRef := vs.Spec.Provider.VirtualNode.VirtualNodeRef
+	vnKey := references.ObjectKeyForVirtualNodeReference(vs, vnRef)
+	vn, err := m.referencesResolver.ResolveVirtualNodeReference(ctx, vs, vnRef)
+	if err != nil {
+		return nil, err
+	}
+	return map[types.NamespacedName]*appmesh.VirtualNode{
+		vnKey: vn,
+	}, nil
+}
+
+func (m *defaultResourceManager) validateVirtualNodeDependencies(ctx context.Context, ms *appmesh.Mesh, vnByKey map[types.NamespacedName]*appmesh.VirtualNode) error {
+	for _, vn := range vnByKey {
+		if vn.Spec.MeshRef == nil || !mesh.IsMeshReferenced(ms, *vn.Spec.MeshRef) {
+			return errors.Errorf("virtualNode %v didn't belong to mesh %v", k8s.NamespacedName(vn), k8s.NamespacedName(ms))
+		}
+		if !virtualnode.IsVirtualNodeActive(vn) {
+			return runtime.NewRequeueError(errors.New("virtualNode is not active yet"))
+		}
+	}
+	return nil
+}
+
+func (m *defaultResourceManager) findVirtualRouterDependencies(ctx context.Context, vs *appmesh.VirtualService) (map[types.NamespacedName]*appmesh.VirtualRouter, error) {
```

Comment:
Nit: Why is this called `findVirtualRouterDependencies` when it can only ever return a single key and value or nil? `vrKey: vr`

Reply:
oh, it's normalized to map[key]value to leverage the shared utilities in referencers package :D